marine_stress.c

```diff
@@ -0,0 +1,110 @@
+#include "marine.h"
+#include "marine_dev.h"
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+
+void fill_random(char *buf, size_t len) {
+    for (size_t i = 0; i < len / sizeof(int); ++i) {
+        *(((int *) buf) + i) = rand();
+    }
+}
+
+void random_eth(char *buf) {
+    fill_random(buf, 12); // randomize macs
+    buf[12] = 0x08;
+    buf[13] = 0;
+}
+
+void random_ip(char *buf) {
+    buf += 14; // skip eth
+    *(buf++) = 0x45;
+    *(buf++) = 0;
+    *(buf++) = 0;
+    *(buf++) = 0x0e;
+    *(buf++) = 0;
+    *(buf++) = 0;
+    *(buf++) = 0;
+    *(buf++) = 0;
+    *(buf++) = 0x40;
+    *(buf++) = 0x06;
+    *(buf++) = 0xac;
+    *(buf++) = 0x16;
+    fill_random(buf, 8); // randomize ips
+}
+
+void random_tcp(char *buf) {
+    buf += 14; // skip eth
+    buf += 20; // skip ip
+    fill_random(buf, 4); // randomize ports
+    buf += 4;
+    *(buf++) = 0xde;
+    *(buf++) = 0xad;
+    *(buf++) = 0xbe;
+    *(buf++) = 0xef;
+    *(buf++) = 0;
+    *(buf++) = 0;
+    *(buf++) = 0;
+    *(buf++) = 0;
+    *(buf++) = 0x50;
+    *(buf++) = 0x02;
+    *(buf++) = 0xff;
+    *(buf++) = 0xff;
+    *(buf++) = 0;
+    *(buf++) = 0;
+    *(buf++) = 0;
+    *(buf++) = 0;
+}
+
+int report_mem(size_t rss) {
+    return printf("MEMORY: %.2lfMB\n", rss / 1024.0 / 1024.0);
+}
+
+size_t report_current_mem(void) {
+    size_t rss = get_current_rss();
+    report_mem(rss);
+    return rss;
+}
+
+#define DATA_LEN 800
+#define CHUNK (1U << 19U)
+#define TOTAL_PACKETS (CHUNK << 3U)
+
+int main(void) {
+    srand(time(NULL));
+    char *fields[] = {"eth.src", "ip.dst", "tcp.srcport"};
+    char err_msg[512] = {0};
+    report_current_mem();
+    printf("Loading marine...\n");
+    init_marine();
+    report_current_mem();
+    printf("Adding filter\n");
+    int filter_id = marine_add_filter("ether[0] & 1 == 0", "frame[0] & 2", fields,
+                                      3, err_msg);
+    if (filter_id < 0) {
+        fprintf(stderr, "Could not add filter: %s\n", err_msg);
+        return -1;
+    }
+    char data[DATA_LEN] = {0};
+    size_t prev_rss;
+    size_t rss = report_current_mem();
+    for (size_t i = 1; i < TOTAL_PACKETS; ++i) {
+        if (i % CHUNK == 0) {
+            prev_rss = rss;
+            rss = report_current_mem();
+            printf("bytes-per-packet: %.2lf\n", ((double) (rss - prev_rss)) / CHUNK);
+        }
+        fill_random(data, DATA_LEN);
+        random_eth(data);
+        unsigned int flags = rand();
+        if ((flags & 1U) == 0) {
```

Comment:
What do we gain from this `if`? Wouldn't it be better to constantly randomize IPs to have more consistent stress tests? (Same applies to tcp)

Reply:
Alternatively, we could use a closed set of conversations

Reply:
I'd like this stress test to also cover weird cases (raw eth packets, raw IP packets).
Generating a closed set of conversations will defeat the purpose, since it won't explode the caches. We want marine to see a lot of non related packets, like in real world uses.

Reply:
The closed set can be huge. 

Maybe you could add the output of a few runs to answer this question, but doesn't random inconsistent results? If it doesn't then why? 

I guess eventually we'll want to add `assert`s to the stress test to make sure everything is fine without having to manually look at output

Reply:
Also another note about the bytes per packet - is it consistent across layers?

For example, maybe TCP on average is +20 per packet, but ethernet is +0 per packet? Maybe by isolating layers we'll be able to pinpoint more caches.